zerver/tests/test_auth_backends.py

```diff
@@ -942,6 +945,151 @@ def test_social_auth_complete_when_base_exc_is_raised(self) -> None:
             self.assertEqual(result.status_code, 302)
             self.assertIn('login', result.url)
 
+class SAMLAuthBackendTest(SocialAuthBase):
+    __unittest_skip__ = False
+
+    BACKEND_CLASS = SAMLAuthBackend
+    CLIENT_KEY_SETTING = "SOCIAL_AUTH_SAML_SP_PRIVATE_KEY"
+    LOGIN_URL = "/accounts/login/social/saml"
+    SIGNUP_URL = "/accounts/register/social/saml"
+    AUTHORIZATION_URL = "https://idp.testshib.org/idp/profile/SAML2/Redirect/SSO"
+    AUTH_FINISH_URL = "/complete/saml/"
+    CONFIG_ERROR_URL = "/config-error/saml"
+
+    # We have to define our own social_auth_test as the flow of SAML authentication
+    # is different from the other social backends.
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         is_signup: Optional[str]=None,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         **extra_data: Any) -> HttpResponse:
+        url = self.LOGIN_URL
+
+        params = {}
+        headers = {}
+        if subdomain is not None:
+            headers['HTTP_HOST'] = subdomain + ".testserver"
+        if mobile_flow_otp is not None:
+            params['mobile_flow_otp'] = mobile_flow_otp
+            headers['HTTP_USER_AGENT'] = "ZulipAndroid"
+        if is_signup is not None:
+            url = self.SIGNUP_URL
+        params['next'] = next
+        params['multiuse_object_key'] = multiuse_object_key
+        if len(params) > 0:
+            url += "?%s" % (urllib.parse.urlencode(params),)
```

Comment:
Is this block identical with the normal `social_auth_test`?  If so I think we should extract this block as a function `self.get_social_auth_http_params()` that returns (url, params, headers) to avoid duplicating this somewhat complex block of code.

Reply:
extracted, as ``prepare_login_url_and_headers`` in a separate commit now.